mlperf_common/callbacks/logging.py

```diff
@@ -0,0 +1,344 @@
+            mllogger.log_init_stop_run_start()
+            self._start_train_block(
+                trainer.global_step,
+                self.get_train_step_samples_count(trainer, pl_module),
+            )
+            self.timer.reset()
+
+    def on_train_end(
+        self,
+        trainer: pl.Trainer,
+        pl_module: pl.LightningModule,
+    ) -> None:
+        if self.train_block_started:
+            self._end_train_block(self.get_train_step_samples_count(trainer, pl_module))
+
+        if hasattr(trainer, "run_stop_logged") and not trainer.run_stop_logged:
+            train_batch_size = self.get_train_step_samples_count(trainer, pl_module)
+            if (
+                hasattr(trainer, "target_reached") and trainer.target_reached
+            ) or self.force_success:
+                status = mllogger.constants.SUCCESS
+            else:
+                status = mllogger.constants.ABORTED
+
+            mllogger.end(
+                mllogger.constants.RUN_STOP,
+                metadata={
+                    mllogger.constants.SAMPLES_COUNT: self.train_current_block
+                    * train_batch_size,
+                    "status": status,
+                },
+            )
+
+    def on_validation_start(
+        self,
+        trainer: pl.Trainer,
+        pl_module: pl.LightningModule,
+    ) -> None:
+        if not trainer.warmup:
+            train_batch_size = self.get_train_step_samples_count(trainer, pl_module)
+            self._log_training_throughput(
+                trainer.global_step,
+                train_batch_size,
+            )
+
+            if self.train_block_started:
+                self._end_train_block(train_batch_size)
+
+            mllogger.start(
+                mllogger.constants.EVAL_START,
+                metadata={
+                    mllogger.constants.SAMPLES_COUNT: trainer.global_step
+                    * train_batch_size
+                },
+            )
+
+    def on_validation_end(
+        self,
+        trainer: pl.Trainer,
+        pl_module: pl.LightningModule,
+    ) -> None:
+        if not trainer.warmup:
+            train_batch_size = self.get_train_step_samples_count(trainer, pl_module)
+            validation_samples = self.get_validation_samples_count(trainer, pl_module)
+
+            self._log_validation_throughput(
+                trainer.global_step,
+                train_batch_size,
+                validation_samples,
+            )
+
+            mllogger.end(
+                mllogger.constants.EVAL_STOP,
+                metadata={
+                    mllogger.constants.SAMPLES_COUNT: trainer.global_step
+                    * train_batch_size
+                },
+            )
+
+            if not trainer.should_stop:
+                self._start_train_block(
+                    trainer.global_step,
+                    train_batch_size,
+                )
+
+    def log_custom_timedelta(self, value_key, step: int = 0):
```

Comment:
In Lora we call this 5 times: three times in custom_callbacks.py (for cuda_graph start and end and warmup), and twice in custom_llama.py (around model init).
I guess all those things are happening outside this source code in some other callbacks file?  Does that callbacks file have access to this function in this file?

Reply:
yes, there is no way to universally handle warmup and cuda graphs for now.
Both lora and llama will build on top of this base callback, filling benchmark specific informations, while general logging will be handled here.

Alternatively, both warmup and cuda_graphs can be handled with different callback, but then it is easy to mess up order of in which each callback is called. Extending generic callback with additional functionality seems to be a less error prone solution

Reply:
And what's the deal with the call to `self.warmup()` on line 68?  If that's actually doing warmup there should probably be calls to `log_custom_timedelta()` around it as well.

Reply:
Added call to `self.log_custom_timedelta()`, with a key "warmup-time" after `self.warmup()` is finished, so it isn't necessary to handle that manually inside warmup